feat(tools): add ToolOutputFilter service — ANSI strip, dedup, success short-circuit, failure extraction

## Problem

AutoBot's agent loop passes raw tool/command stdout back to the LLM with only naive byte-limit truncation at four scattered, uncoordinated points:

| File | Current approach | Problem |
|------|-----------------|---------|
| \`tools/code_interpreter.py:67\` | \`raw_stdout[:10_240]\` | Blind cut from end — loses failures at the bottom |
| \`chat_workflow/manager.py:1687\` | \`output_text[:2000]\` | Same blind cut |
| \`api/git_mcp.py:302\` | \`stdout_str[:1_048_576]\` | 1 MB cap, no semantic filtering |
| \`secure_command_executor.py\` | No size limit | Raw ANSI escape codes reach the LLM |

There is no pipeline that removes noise before output reaches the LLM. Every \`git push\`, \`pytest\` run, \`docker pull\`, and \`npm install\` sends multi-line boilerplate the LLM does not need to read.

---

## Proposed Solution

Introduce a \`ToolOutputFilter\` service wired into \`agent_loop/loop.py\` (between tool execution and \`tool_result\` message assembly). The filter applies a 15-method pipeline to every command output before it reaches the LLM.

### Pipeline (applied in order)

```
1. strip_ansi            ← always, free
2. apply_no_op_detection ← if returns non-None, stop here  
3. short_circuit_write   ← git/docker write ops → "ok branch"
4. route to command filter:
   - pytest  → state-machine filter
   - ruff    → JSON grouping
   - git diff → diff condenser
   - gh body  → markdown filter
   - docker logs → consecutive dedup
   - default  → tail_lines(100)   ← last N not first N
5. join_with_overflow on any lists
6. tee_and_hint          ← save raw to disk, append path hint
```

---

## Method Specifications

### Method 1 — ANSI Stripping

Applied first, always. Single compiled regex.

```python
import re
_ANSI_RE = re.compile(r'\x1b\[[0-9;]*[a-zA-Z]')

def strip_ansi(text: str) -> str:
    return _ANSI_RE.sub('', text)
```

**Wire into:** all stdout/stderr paths — \`code_interpreter.py\`, \`secure_command_executor.py\`, \`git_mcp.py\`, \`secure_sandbox_executor.py\`.

---

### Method 2 — Pytest State Machine

4-state machine: `Header → Progress → Failures → Summary`. Collects the full `=== FAILURES ===` section; discards everything else (passing test lines, platform info, progress dots). Handles both verbose and `-q` mode (quiet mode has no `===` wrapper on the summary line — requires explicit detection).

```python
from enum import Enum, auto

class _PytestState(Enum):
    HEADER = auto(); PROGRESS = auto(); FAILURES = auto(); SUMMARY = auto()

def filter_pytest(output: str, exit_code: int) -> str:
    state = _PytestState.HEADER
    failures: list[list[str]] = []
    current: list[str] = []
    summary = ""

    for line in output.splitlines():
        t = line.strip()
        if "=== test session starts ===" in t:
            state = _PytestState.HEADER
        elif "=== FAILURES ===" in t or "=== ERRORS ===" in t:
            state = _PytestState.FAILURES
        elif "=== short test summary" in t:
            if current: failures.append(current); current = []
            state = _PytestState.SUMMARY
        elif t.startswith("===") and any(w in t for w in ("passed","failed","skipped","error")):
            summary = t.strip("= ")
        # quiet-mode bare summary (no === wrapper)
        elif (not summary and not t.startswith("===")
              and not t.startswith(("FAILED","ERROR"))
              and any(w in t for w in (" passed"," failed"," skipped"))
              and " in " in t):
            summary = t
        elif state == _PytestState.HEADER and t.startswith("collected"):
            state = _PytestState.PROGRESS
        elif state == _PytestState.FAILURES:
            if t.startswith("___"):
                if current: failures.append(current); current = []
                current = [t]
            elif t and not t.startswith("==="):
                current.append(t)
        elif state == _PytestState.SUMMARY:
            if t.startswith(("FAILED","ERROR")):
                failures.append([t])

    if current: failures.append(current)

    passed, failed, skipped = _parse_pytest_summary(summary)
    if not failures and exit_code == 0:
        return f"pytest: {passed} passed" + (f", {skipped} skipped" if skipped else "")

    header = f"pytest: {passed} passed, {failed} failed"
    if skipped: header += f", {skipped} skipped"
    parts = [header, "─" * 40]
    for block in failures[:5]:
        parts.append(_format_failure_block(block))
    if len(failures) > 5:
        parts.append(f"... +{len(failures)-5} more failures")
    return "\n".join(parts)
```

**Wire into:** \`tools/code_interpreter.py\`, \`secure_command_executor.py\` for any command matching \`^pytest\` or \`^python -m pytest\`.

---

### Method 3 — JSON-Mode Linter Parsing (Ruff)

Two parts: (a) inject \`--output-format=json\` before execution, (b) parse JSON and group by rule code + file.

```python
def filter_ruff_json(stdout: str) -> str:
    try:
        diagnostics = json.loads(stdout)
    except json.JSONDecodeError:
        return stdout  # passthrough

    if not diagnostics:
        return "ruff: no issues found"

    by_rule: dict[str, int] = {}
    by_file: dict[str, int] = {}
    fixable = 0
    for d in diagnostics:
        by_rule[d["code"]] = by_rule.get(d["code"], 0) + 1
        by_file[d["filename"]] = by_file.get(d["filename"], 0) + 1
        if d.get("fix"):
            fixable += 1

    top_rules = sorted(by_rule.items(), key=lambda x: -x[1])[:10]
    top_files = sorted(by_file.items(), key=lambda x: -x[1])[:10]

    lines = [
        f"ruff: {len(diagnostics)} issues in {len(by_file)} files"
        + (f" ({fixable} fixable)" if fixable else ""),
        "─" * 40,
        "Top rules: " + ", ".join(f"{r}({n})" for r, n in top_rules),
    ]
    for f, n in top_files:
        lines.append(f"  {_compact_path(f)} ({n})")
    if fixable:
        lines.append(f"\n[hint] ruff check --fix to auto-fix {fixable}")
    return "\n".join(lines)

def _compact_path(path: str) -> str:
    for prefix in ("/src/", "/lib/", "/tests/", "/autobot-backend/"):
        if prefix in path:
            return path[path.index(prefix)+1:]
    return path.rsplit("/", 1)[-1]
```

**Flag injection point** (before dispatch):
```python
if cmd.startswith("ruff check") and "--output-format" not in cmd:
    cmd = cmd.replace("ruff check", "ruff check --output-format=json", 1)
```

---

### Method 4 — Exit-Code-Aware Write Short-Circuit

Combines exit code + output pattern + command type. Never short-circuits on non-zero exit.

```python
import re

_GIT_PUSH_BRANCH = re.compile(r"([\w/.-]+)\s+->\s+\w+/([\w/.-]+)")
_GIT_COMMIT_HASH = re.compile(r"\[[\w/.-]+\s+([a-f0-9]{7,})\]")

def short_circuit_git(subcmd: str, stdout: str, stderr: str, exit_code: int) -> str | None:
    if exit_code != 0:
        return None
    if subcmd in ("add", "rm"):
        return "ok" if not stderr.strip() else None
    elif subcmd == "commit":
        m = _GIT_COMMIT_HASH.search(stdout)
        return f"ok {m.group(1)}" if m else "ok"
    elif subcmd == "push":
        if "Everything up-to-date" in stdout:
            return "ok (already up-to-date)"
        m = _GIT_PUSH_BRANCH.search(stdout)
        return f"ok {m.group(2)}" if m else "ok"
    elif subcmd == "fetch":
        return "ok (nothing new)" if not stdout.strip() and not stderr.strip() else None
    return None
```

**Wire into:** \`api/git_mcp.py\` per-command handlers, \`agent_loop/loop.py\` after \`_dispatch_tool()\`.

---

### Method 5 — Block Stream Filter

A `BlockHandler` protocol for tools that produce structured error blocks (ESLint, mypy, docker build). Accumulates lines that are continuations (indented) of a block-start line.

```python
from typing import Protocol

class BlockHandler(Protocol):
    def should_skip(self, line: str) -> bool: ...
    def is_block_start(self, line: str) -> bool: ...
    def is_block_continuation(self, line: str, block: list[str]) -> bool: ...
    def format_summary(self, exit_code: int, blocks: list[list[str]]) -> str: ...

def filter_by_blocks(output: str, handler: BlockHandler, exit_code: int) -> str:
    blocks: list[list[str]] = []
    current: list[str] = []
    for line in output.splitlines():
        if handler.should_skip(line):
            continue
        if handler.is_block_start(line):
            if current: blocks.append(current)
            current = [line]
        elif current and handler.is_block_continuation(line, current):
            current.append(line)
        else:
            if current: blocks.append(current); current = []
    if current: blocks.append(current)
    return handler.format_summary(exit_code, blocks)
```

**Use cases:** ESLint output (block starts with file path, continuations are rule violations), mypy (block starts with `filename:line:`, continuations are context), docker build failures.

---

### Method 6 — Consecutive Line Deduplication

Collapses runs of identical lines into `"line [×N]"`. Critical for docker logs and container crash loops.

```python
def dedup_consecutive(lines: list[str], min_repeat: int = 3) -> list[str]:
    if not lines:
        return lines
    result = []
    prev, count = lines[0], 1
    for line in lines[1:]:
        if line == prev:
            count += 1
        else:
            result.append(f"{prev} [×{count}]" if count >= min_repeat else "\n".join([prev]*count))
            prev, count = line, 1
    result.append(f"{prev} [×{count}]" if count >= min_repeat else prev)
    return result
```

**Wire into:** \`secure_sandbox_executor.py:220\` (container logs), any long-running process stdout.

---

### Method 7 — `join_with_overflow` (Semantic List Truncation)

Cuts at item boundaries, not byte boundaries. Tells the LLM exactly how much was omitted.

```python
def join_with_overflow(items: list[str], max_items: int, label: str = "items") -> str:
    result = "\n".join(items[:max_items])
    if len(items) > max_items:
        result += f"\n... +{len(items)-max_items} more {label}"
    return result
```

**Apply to:** file lists in git status, error lists in mypy output, any `tool_result` list before sending to LLM.

---

### Method 8 — `tail_lines` Fallback (Last N, Not First N)

Failures are at the **end** of tool output. Current \`output[:10_240]\` keeps the beginning (platform info, setup lines). Switch to tail as the default fallback.

```python
def tail_lines(output: str, n: int) -> str:
    lines = output.splitlines()
    if len(lines) <= n:
        return output
    omitted = len(lines) - n
    return f"[... {omitted} lines omitted ...]\n" + "\n".join(lines[-n:])
```

**Replace:** \`raw_stdout[:MAX_OUTPUT_BYTES]\` in \`code_interpreter.py:68\`, \`output_text[:2000]\` in \`manager.py:1687\`.

---

### Method 9 — Flag Injection (Compact Output at Source)

Modify the command before execution to produce smaller, more structured output.

```python
_FLAG_INJECTIONS: list[tuple[str, str, str]] = [
    # (match_prefix, flag_to_inject, guard_substring)
    ("pytest",      "--tb=short -q",          "--tb"),
    ("ruff check",  "--output-format=json",    "--output-format"),
    ("go test",     "-json",                   "-json"),
]

def inject_compact_flags(command: str) -> str:
    for prefix, flag, guard in _FLAG_INJECTIONS:
        if command.startswith(prefix) and guard not in command:
            return command.replace(prefix, f"{prefix} {flag}", 1)
    return command
```

**Wire into:** tool dispatch point, before command is executed (not at filter time).

---

### Method 10 — No-Op Detection

Detects clean-state messages and collapses them to "ok". Exit-code-aware: only fires on exit_code == 0.

```python
import re

_NO_OP_PATTERNS: dict[str, list[tuple[re.Pattern, str]]] = {
    "git status":  [(re.compile(r"nothing to commit.*working tree clean"), "clean"),
                    (re.compile(r"nothing added to commit"), "unstaged changes (use git add)")],
    "git pull":    [(re.compile(r"Already up to date"), "ok (already up-to-date)")],
    "docker pull": [(re.compile(r"Status: Image is up to date"), "ok (already up-to-date)")],
    "pip install": [(re.compile(r"Requirement already satisfied"), "ok (already installed)")],
}

def apply_no_op_detection(command: str, stdout: str, exit_code: int) -> str | None:
    if exit_code != 0:
        return None
    if not stdout.strip():
        return "ok"
    for prefix, patterns in _NO_OP_PATTERNS.items():
        if command.startswith(prefix):
            for pattern, message in patterns:
                if pattern.search(stdout):
                    return message
    return None
```

---

### Method 11 — Tee (Save Full Output, Append File Hint)

When output is filtered/truncated: save raw output to a timestamped file, append path hint so the LLM has a fallback read path. Three modes: `failures` (default), `always`, `never`. Auto-rotates to keep last 20 files. Skips outputs < 500 bytes.

```python
import os, time, pathlib

_TEE_DIR = pathlib.Path.home() / ".local/share/autobot/tee"
_MAX_FILES = 20
_MIN_BYTES = 500

def tee_and_hint(raw: str, slug: str, exit_code: int, mode: str = "failures") -> str | None:
    if mode == "never" or len(raw.encode()) < _MIN_BYTES:
        return None
    if mode == "failures" and exit_code == 0:
        return None
    _TEE_DIR.mkdir(parents=True, exist_ok=True)
    safe_slug = "".join(c if c.isalnum() or c in "-_" else "_" for c in slug)[:40]
    path = _TEE_DIR / f"{int(time.time())}_{safe_slug}.log"
    path.write_text(raw, encoding="utf-8")
    for old in sorted(_TEE_DIR.glob("*.log"))[:-_MAX_FILES]:
        old.unlink(missing_ok=True)
    hint = str(path).replace(str(pathlib.Path.home()), "~")
    return f"[full output: {hint}]"
```

**Integration in \`ToolOutputFilter.apply()\`:**
```python
filtered = _route_to_filter(command, stdout, exit_code)
if len(stdout) > len(filtered) + 100:
    hint = tee_and_hint(stdout, command.split()[0], exit_code)
    if hint:
        filtered += f"\n{hint}"
```

---

### Method 12 — Diff Condensing

Strips all unified diff metadata (`diff --git`, `--- a/`, `+++ b/`, `@@ ... @@`). Groups `+`/`-` lines by file with a header. **Never truncates diff content** — only metadata is removed. Context lines (no prefix) are dropped.

```python
def condense_unified_diff(diff: str, max_changes_per_file: int = 30) -> str:
    sections: list[str] = []
    current_file = ""
    adds = removes = 0
    changes: list[str] = []

    def flush():
        if not current_file or not changes:
            return
        header = f"[file] {current_file} (+{adds} -{removes})"
        shown = changes[:max_changes_per_file]
        body = "\n".join(f"  {c}" for c in shown)
        overflow = len(changes) - max_changes_per_file
        tail = f"\n  ... +{overflow} more" if overflow > 0 else ""
        sections.append(f"{header}\n{body}{tail}")

    for line in diff.splitlines():
        if line.startswith("+++ b/"):
            flush()
            current_file = line[6:]; adds = removes = 0; changes = []
        elif line.startswith(("diff --git", "--- ", "+++ ", "@@")):
            continue
        elif line.startswith("+") and not line.startswith("+++"):
            adds += 1; changes.append(line)
        elif line.startswith("-") and not line.startswith("---"):
            removes += 1; changes.append(line)

    flush()
    return "\n".join(sections) if sections else "[no diff]"
```

**Wire into:** \`api/git_mcp.py\` git diff handler, replacing the 1 MB naive truncation.

---

### Method 13 — Markdown Body Filtering

For GitHub PR/issue body text: strip HTML comments, badge lines, image-only lines, horizontal rules, collapse multi-blank-lines. Preserves code blocks (`\`\`\`` / `~~~` fences) untouched.

```python
import re

_HTML_COMMENT = re.compile(r'<!--.*?-->', re.DOTALL)
_BADGE_LINE   = re.compile(r'^\s*\[!\[[^\]]*\]\([^)]*\)\]\([^)]*\)\s*$', re.MULTILINE)
_IMAGE_LINE   = re.compile(r'^\s*!\[[^\]]*\]\([^)]*\)\s*$', re.MULTILINE)
_HR_LINE      = re.compile(r'^\s*(?:-{3,}|\*{3,}|_{3,})\s*$', re.MULTILINE)
_MULTI_BLANK  = re.compile(r'\n{3,}')

def filter_markdown_body(text: str) -> str:
    result = []
    remaining = text
    while remaining:
        candidates = [(remaining.find(f), f) for f in ("```", "~~~") if remaining.find(f) != -1]
        if candidates:
            pos, fence = min(candidates)
            result.append(_filter_md_segment(remaining[:pos]))
            after_open = remaining.find('\n', pos + len(fence))
            if after_open == -1: result.append(remaining[pos:]); break
            close = remaining.find(fence, after_open + 1)
            if close == -1: result.append(remaining[pos:]); break
            close_end = (remaining.find('\n', close) + 1) or len(remaining)
            result.append(remaining[pos:close_end])
            remaining = remaining[close_end:]
        else:
            result.append(_filter_md_segment(remaining)); break
    return _MULTI_BLANK.sub('\n\n', ''.join(result)).strip()

def _filter_md_segment(text: str) -> str:
    for pattern in (_HTML_COMMENT, _BADGE_LINE, _IMAGE_LINE, _HR_LINE):
        text = pattern.sub('', text)
    return text
```

**Wire into:** anywhere AutoBot reads GitHub issue/PR bodies — \`api/git_mcp.py\` \`gh issue view\` / \`gh pr view\` handlers, agent loop issue-reading paths.

---

### Method 14 — Similarity-Based Line Classification (Diff)

Jaccard character-set overlap to distinguish a line "modification" from a "deletion + addition". Used inside diff condensing to produce `~N old → new` format for similar lines.

```python
def _line_similarity(a: str, b: str) -> float:
    sa, sb = set(a), set(b)
    union = sa | sb
    return len(sa & sb) / len(union) if union else 1.0

# In condense_unified_diff, for adjacent remove+add pairs:
# if _line_similarity(removed, added) > 0.5:
#     emit f"~{line_no}  {removed} → {added}"
# else:
#     emit remove and add separately
```

**Apply to:** \`condense_unified_diff\` when processing paired remove+add lines at the same line number.

---

### Method 15 — Command Classification + Savings Tracking

After filtering, record `(tool_name, category, original_bytes, filtered_bytes)` to Redis. Enables discovering which tool types burn the most context tokens per session.

```python
import re

_TOOL_CATEGORIES: list[tuple[re.Pattern, str, int]] = [
    (re.compile(r"^pytest"),                  "Tests",     800),
    (re.compile(r"^ruff"),                    "Lint",      100),
    (re.compile(r"^git (log|diff|show)"),     "Git-read",  200),
    (re.compile(r"^git (add|commit|push|pull|fetch)"), "Git-write", 40),
    (re.compile(r"^gh "),                     "GitHub",    200),
    (re.compile(r"^docker"),                  "Infra",     120),
    (re.compile(r"^npm|^pnpm"),               "JS",        300),
]

def classify_tool(command: str) -> tuple[str, int]:
    for pattern, category, avg_tokens in _TOOL_CATEGORIES:
        if pattern.match(command):
            return category, avg_tokens
    return "Other", 150

async def record_filter_savings(command: str, original: str, filtered: str) -> None:
    category, _ = classify_tool(command)
    saved = len(original) - len(filtered)
    if saved <= 0:
        return
    redis = await get_async_redis_client()
    if redis is None:
        return
    key = f"tool_filter_savings:{category}:{datetime.utcnow().date()}"
    await redis.incrby(key, saved)
    await redis.expire(key, 90 * 86400)  # 90-day retention
```

Query with \`rtk gain\`-equivalent: sum bytes saved per category over last N days.

---

## Acceptance Criteria

- [ ] \`autobot-backend/services/tool_output_filter.py\` — \`ToolOutputFilter\` service with all 15 methods
- [ ] \`autobot-backend/config/tool_output_filters.yaml\` — declarative per-command rules
- [ ] Method 1 (ANSI) wired into all 4 existing stdout paths
- [ ] Methods 2–4 (pytest, ruff, git write) routing in place
- [ ] Method 5 (BlockFilter) implemented and usable by ESLint/mypy handlers
- [ ] Method 6 (dedup) wired into sandbox executor container logs
- [ ] Methods 7–8 (join_with_overflow, tail_lines) replacing all naive byte-slice truncation
- [ ] Method 9 (flag injection) wired at tool dispatch for pytest and ruff
- [ ] Method 10 (no-op detection) covers git status/pull, docker pull, pip install
- [ ] Method 11 (tee) wired in \`ToolOutputFilter.apply()\` on meaningful compression
- [ ] Method 12 (diff condenser) replacing naive truncation in \`api/git_mcp.py\` git diff handler
- [ ] Method 13 (markdown filter) applied to all \`gh issue view\` / \`gh pr view\` body reads
- [ ] Method 14 (similarity) used inside diff condenser for paired remove+add lines
- [ ] Method 15 (classifier + Redis savings tracking) recording per-category bytes saved
- [ ] Unit tests for each method independently
- [ ] No behaviour change for unmatched commands (passthrough)

## Files Affected

- \`autobot-backend/services/tool_output_filter.py\` (new)
- \`autobot-backend/config/tool_output_filters.yaml\` (new)
- \`autobot-backend/agent_loop/loop.py\`
- \`autobot-backend/tools/code_interpreter.py\`
- \`autobot-backend/chat_workflow/manager.py\`
- \`autobot-backend/api/git_mcp.py\`
- \`autobot-backend/secure_command_executor.py\`
- \`autobot-backend/secure_sandbox_executor.py\`
- \`autobot-backend/tests/services/test_tool_output_filter.py\` (new)

--- a/autobot-backend/api/git_mcp.py
+++ b/autobot-backend/api/git_mcp.py
@@ -47,7 +47,10 @@ from autobot_shared.security.path_validator import validate_path
 from autobot_shared.ssot_config import PROJECT_ROOT
 from autobot_shared.time_utils import now_utc
 from constants.threshold_constants import QueryDefaults
+from services.tool_output_filter import ToolOutputFilter
 from type_defs.common import JSONObject, Metadata
+
+_output_filter = ToolOutputFilter()
 
 logger = logging.getLogger(__name__)
 
@@ -298,9 +301,10 @@ async def _run_git_process(cmd: List[str], repo_path: str, timeout: int) -> Meta
     stdout_str = stdout.decode("utf-8", errors="replace")
     stderr_str = stderr.decode("utf-8", errors="replace")
 
-    # Check output size
+    # Check output size, then apply semantic filter
     if len(stdout_str) > MAX_OUTPUT_SIZE:
-        stdout_str = stdout_str[:MAX_OUTPUT_SIZE] + "\n... (output truncated)"
+        stdout_str = stdout_str[:MAX_OUTPUT_SIZE]
+    stdout_str = _output_filter.filter(" ".join(cmd[:3]), stdout_str)
 
     return {
         "success": process.returncode == 0,

--- a/autobot-backend/chat_workflow/manager.py
+++ b/autobot-backend/chat_workflow/manager.py
@@ -30,8 +30,10 @@ from .llm_handler import LLMHandlerMixin, _emit_after_continuation, _emit_before
 from .models import LLMIterationContext, StreamingMessage, WorkflowSession
 from .session_handler import SessionHandlerMixin, _emit_approval_received, _emit_approval_required
 from .tool_handler import ToolHandlerMixin
+from services.tool_output_filter import ToolOutputFilter
 
 logger = logging.getLogger(__name__)
+_output_filter = ToolOutputFilter()
 
 # Issue #380: Module-level frozenset for terminal message types
 _TERMINAL_MESSAGE_TYPES: FrozenSet[str] = frozenset(
@@ -1682,10 +1684,7 @@ class ChatWorkflowManager(
         if stderr:
             output_text += f"\nStderr: {stderr}"
 
-        # Issue #650: Increased from 500 to 2000 for better continuation context
-        max_output_len = 2000
-        if len(output_text) > max_output_len:
-            output_text = output_text[:max_output_len] + "\n... (output truncated)"
+        output_text = _output_filter.filter(cmd, output_text)
 
         return f"**Step {step_num}:** `{cmd}`\n- Status: {status}\n- Output:\n```\n{output_text}\n```"
 

--- a/autobot-backend/config/tool_output_filters.yaml
+++ b/autobot-backend/config/tool_output_filters.yaml
@@ -1,0 +1,67 @@
+# Tool output filter rules for ToolOutputFilter service.
+# Each key under `filters:` is a named rule; first match on match_command wins.
+#
+# Pipeline stages (applied in order when present):
+#   strip_ansi           — remove ANSI/VT100 escape codes
+#   match_output         — if output matches pattern, replace with short message
+#   strip_lines_matching — drop lines matching any regex
+#   keep_lines_matching  — keep only lines matching any regex
+#   dedup_consecutive    — collapse repeated consecutive lines into "line [×N]"
+#   max_lines            — keep last N lines
+#   on_empty             — message returned when filtered result is empty
+
+filters:
+  git_write:
+    match_command: "^git (add|push|pull|commit|fetch|merge|rebase|stash)"
+    strip_ansi: true
+    match_output:
+      - pattern: "Everything up-to-date"
+        message: "ok (already up-to-date)"
+      - pattern: "nothing to commit"
+        message: "ok (nothing to commit)"
+    max_lines: 15
+
+  pytest:
+    match_command: "^pytest"
+    strip_ansi: true
+    strip_lines_matching:
+      - "^test_.* PASSED"
+      - "^\\."
+      - "^s$"
+      - "^platform "
+      - "^rootdir:"
+      - "^plugins:"
+      - "^collecting \\.\\.\\."
+    max_lines: 150
+    on_empty: "All tests passed"
+
+  npm_test:
+    match_command: "^npm (test|run test|run lint|run build|run type-check)"
+    strip_ansi: true
+    strip_lines_matching:
+      - "^  ✓ "
+      - "^  ✔ "
+      - "^  pass "
+    max_lines: 150
+
+  ruff_eslint:
+    match_command: "^(ruff|eslint|flake8|mypy|black)"
+    strip_ansi: true
+    max_lines: 100
+
+  docker_logs:
+    match_command: "^docker (logs|compose logs)"
+    strip_ansi: true
+    dedup_consecutive: true
+    max_lines: 60
+
+  pip_install:
+    match_command: "^pip(3)? install"
+    strip_ansi: true
+    strip_lines_matching:
+      - "^Requirement already satisfied"
+      - "^Collecting "
+      - "^  Downloading "
+      - "^  Using cached "
+    max_lines: 20
+    on_empty: "ok (all requirements already satisfied)"

--- a/autobot-backend/services/tool_output_filter.py
+++ b/autobot-backend/services/tool_output_filter.py
@@ -1,0 +1,116 @@
+# AutoBot - AI-Powered Automation Platform
+# Copyright (c) 2025 mrveiss
+# Author: mrveiss
+"""
+ToolOutputFilter — declarative pipeline for cleaning command stdout before it
+reaches the LLM.  Replaces naive byte-slice truncation at scattered sites.
+
+Usage::
+
+    from services.tool_output_filter import ToolOutputFilter
+    _filter = ToolOutputFilter()
+    clean = _filter.filter("pytest tests/", raw_output)
+"""
+from __future__ import annotations
+
+import logging
+import os
+import re
+from typing import Any
+
+import yaml
+
+logger = logging.getLogger(__name__)
+
+_DEFAULT_CONFIG = os.path.join(
+    os.path.dirname(__file__), "..", "config", "tool_output_filters.yaml"
+)
+
+
+class ToolOutputFilter:
+    """Apply a rule-based pipeline to tool command output."""
+
+    def __init__(self, config_path: str | None = None) -> None:
+        path = config_path or _DEFAULT_CONFIG
+        try:
+            with open(path, encoding="utf-8") as fh:
+                cfg = yaml.safe_load(fh) or {}
+            self._rules: list[dict[str, Any]] = list(cfg.get("filters", {}).values())
+        except FileNotFoundError:
+            logger.warning("tool_output_filter: config not found at %s", path)
+            self._rules = []
+        except Exception:
+            logger.exception("tool_output_filter: failed to load config from %s", path)
+            self._rules = []
+
+    def filter(self, command: str, output: str) -> str:
+        """Return filtered *output* for *command*.  Passthrough if no rule matches."""
+        rule = self._match_rule(command)
+        if rule is None:
+            return output
+        return self._apply(rule, output)
+
+    def _match_rule(self, command: str) -> dict[str, Any] | None:
+        cmd = command.strip()
+        for rule in self._rules:
+            pattern = rule.get("match_command", "")
+            if pattern and re.search(pattern, cmd):
+                return rule
+        return None
+
+    def _apply(self, rule: dict[str, Any], text: str) -> str:
+        if rule.get("strip_ansi"):
+            text = _strip_ansi(text)
+
+        if "match_output" in rule:
+            for entry in rule["match_output"]:
+                if re.search(entry["pattern"], text, re.MULTILINE):
+                    return entry["message"]
+
+        if "strip_lines_matching" in rule:
+            patterns = [re.compile(p) for p in rule["strip_lines_matching"]]
+            text = "\n".join(
+                line for line in text.splitlines()
+                if not any(p.search(line) for p in patterns)
+            )
+
+        if "keep_lines_matching" in rule:
+            patterns = [re.compile(p) for p in rule["keep_lines_matching"]]
+            text = "\n".join(
+                line for line in text.splitlines()
+                if any(p.search(line) for p in patterns)
+            )
+
+        if rule.get("dedup_consecutive"):
+            text = _dedup_consecutive(text)
+
+        max_lines = rule.get("max_lines")
+        if max_lines and max_lines > 0:
+            lines = text.splitlines()
+            if len(lines) > max_lines:
+                text = "\n".join(lines[-max_lines:])
+
+        if not text.strip():
+            return rule.get("on_empty", "")
+
+        return text
+
+
+_ANSI_RE = re.compile(r"\x1b\[[0-9;]*[mGKHF]")
+
+
+def _strip_ansi(text: str) -> str:
+    return _ANSI_RE.sub("", text)
+
+
+def _dedup_consecutive(text: str) -> str:
+    lines = text.splitlines()
+    result: list[str] = []
+    count = 1
+    for i, line in enumerate(lines):
+        if i + 1 < len(lines) and lines[i + 1] == line:
+            count += 1
+        else:
+            result.append(f"{line} [×{count}]" if count > 1 else line)
+            count = 1
+    return "\n".join(result)

--- a/autobot-backend/tests/services/test_tool_output_filter.py
+++ b/autobot-backend/tests/services/test_tool_output_filter.py
@@ -1,0 +1,109 @@
+# AutoBot - AI-Powered Automation Platform
+# Copyright (c) 2025 mrveiss
+# Author: mrveiss
+"""Unit tests for ToolOutputFilter pipeline stages."""
+import os
+import tempfile
+
+import yaml
+
+from services.tool_output_filter import ToolOutputFilter, _strip_ansi, _dedup_consecutive
+
+
+def _make_filter(rules: dict) -> ToolOutputFilter:
+    with tempfile.NamedTemporaryFile(
+        mode="w", suffix=".yaml", delete=False, encoding="utf-8"
+    ) as fh:
+        yaml.dump({"filters": rules}, fh)
+        path = fh.name
+    try:
+        return ToolOutputFilter(config_path=path)
+    finally:
+        os.unlink(path)
+
+
+def test_strip_ansi_removes_color_codes():
+    assert _strip_ansi("\x1b[31mRED\x1b[0m") == "RED"
+
+
+def test_strip_ansi_passthrough_clean():
+    assert _strip_ansi("hello world") == "hello world"
+
+
+def test_dedup_consecutive_collapses():
+    assert _dedup_consecutive("a\na\na\nb\nb\nc") == "a [×3]\nb [×2]\nc"
+
+
+def test_dedup_consecutive_no_repeats():
+    assert _dedup_consecutive("a\nb\nc") == "a\nb\nc"
+
+
+def test_passthrough_when_no_rule():
+    f = _make_filter({})
+    output = "some\noutput\nhere"
+    assert f.filter("cat file.txt", output) == output
+
+
+def test_strip_ansi_stage():
+    f = _make_filter({"r": {"match_command": "^cmd", "strip_ansi": True}})
+    assert f.filter("cmd arg", "\x1b[32mgreen\x1b[0m text") == "green text"
+
+
+def test_match_output_short_circuit():
+    f = _make_filter({
+        "r": {"match_command": "^git", "match_output": [{"pattern": "Everything up-to-date", "message": "ok"}]},
+    })
+    assert f.filter("git push", "Everything up-to-date\nmore stuff") == "ok"
+
+
+def test_match_output_no_match_continues():
+    f = _make_filter({
+        "r": {"match_command": "^git", "match_output": [{"pattern": "Everything up-to-date", "message": "ok"}], "max_lines": 5},
+    })
+    assert "branch pushed" in f.filter("git push", "branch pushed\nremote updated")
+
+
+def test_strip_lines_matching():
+    f = _make_filter({"r": {"match_command": "^pytest", "strip_lines_matching": ["^test_.* PASSED"]}})
+    result = f.filter("pytest tests/", "test_foo PASSED\ntest_bar FAILED\ntest_baz PASSED")
+    assert "PASSED" not in result
+    assert "test_bar FAILED" in result
+
+
+def test_keep_lines_matching():
+    f = _make_filter({"r": {"match_command": "^ruff", "keep_lines_matching": ["error"]}})
+    result = f.filter("ruff check .", "info: checking\nerror: bad code L10\ninfo: done")
+    assert result.strip() == "error: bad code L10"
+
+
+def test_dedup_consecutive_stage():
+    f = _make_filter({"r": {"match_command": "^docker", "dedup_consecutive": True}})
+    result = f.filter("docker logs myapp", "log line\nlog line\nlog line\ndifferent")
+    assert "log line [×3]" in result
+    assert "different" in result
+
+
+def test_max_lines_truncates_to_last_n():
+    f = _make_filter({"r": {"match_command": "^cmd", "max_lines": 3}})
+    assert f.filter("cmd", "\n".join(str(i) for i in range(10))) == "7\n8\n9"
+
+
+def test_on_empty_returned_when_filtered_result_empty():
+    f = _make_filter({"r": {"match_command": "^pytest", "strip_lines_matching": [".*"], "on_empty": "All tests passed"}})
+    assert f.filter("pytest tests/", "test_foo PASSED\ntest_bar PASSED") == "All tests passed"
+
+
+def test_real_config_loads():
+    f = ToolOutputFilter()
+    assert isinstance(f._rules, list)
+
+
+def test_real_config_git_rule_matches():
+    f = ToolOutputFilter()
+    assert "already up-to-date" in f.filter("git push origin main", "Everything up-to-date")
+
+
+def test_real_config_passthrough_unknown_command():
+    f = ToolOutputFilter()
+    output = "hello\nworld"
+    assert f.filter("my_custom_script.sh", output) == output

--- a/autobot-backend/tools/code_interpreter.py
+++ b/autobot-backend/tools/code_interpreter.py
@@ -22,9 +22,12 @@ import sys
 import tempfile
 from typing import Dict, Any
 
+from services.tool_output_filter import ToolOutputFilter
+
 logger = logging.getLogger(__name__)
 
 MAX_OUTPUT_BYTES = 10 * 1024  # 10 KB per stream
+_output_filter = ToolOutputFilter()
 
 
 def execute_code(code: str, timeout_seconds: int = 30) -> Dict[str, Any]:
@@ -64,8 +67,11 @@ def execute_code(code: str, timeout_seconds: int = 30) -> Dict[str, Any]:
             len(raw_stdout) > MAX_OUTPUT_BYTES or len(raw_stderr) > MAX_OUTPUT_BYTES
         )
 
+        stdout_text = _output_filter.filter(
+            "python", raw_stdout[:MAX_OUTPUT_BYTES].decode("utf-8", errors="replace")
+        )
         return {
-            "stdout": raw_stdout[:MAX_OUTPUT_BYTES].decode("utf-8", errors="replace"),
+            "stdout": stdout_text,
             "stderr": raw_stderr[:MAX_OUTPUT_BYTES].decode("utf-8", errors="replace"),
             "exit_code": result.returncode,
             "truncated": truncated,

